lib/wallet/src/Cardano/Wallet/Read/ProtocolParameters.hs

```diff
@@ -0,0 +1,45 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Copyright: © 2023 IOHK
+-- License: Apache-2.0
+--
+-- Provides protocol parameters.
+--
+module Cardano.Wallet.Read.ProtocolParameters
```

Comment:
Shouldn't this be in `Write`?

Reply:
> Shouldn't this be in `Write`?

Since protocol parameters are generally "read" from the blockchain, I simply chose to locate the `ProtocolParameters` type within the `Read` module hierarchy instead of the `Write` hierarchy.

Reply:
I should add that this version of the `ProtocolParameters` is not in any way intended to be the final version of the type we would ultimately like to see in the `Read`/`Write` hierarchy.

It's merely an attempt to solve our current performance issue.

In follow-up PRs, we are still free to do any or all of the following:
- move the type somewhere else
- make the type abstract
- change the type definition

Reply:
> Since protocol parameters are generally "read" from the blockchain, 

Yes. As the purpose of the `Cardano.Wallet.Read` hierarchy is to provide Haskell types and functions for stuff that appears on the blockchain, it is within the purview of the `Read` hierarchy to contain a module for `ProtocolParameters`. This can be the beginning of that module.

In contrast, the purpose of the `Cardano.Wallet.Write` hierarchy is to provide Haskell types and functions for stuff that is accepted into the current and subsequent era of the blockchain. Sometimes, no equivalent in `Read` is needed, but often, `Write` is a selected subset of `Read`. Hence, a typical case is two modules:

* `Cardano.Wallet.Read.ProtocolParameters`
* `Cardano.Wallet.Write.ProtocolParameters` — re-exporting a subset of the `Cardano.Wallet.Read.ProtocolParameters` with specialized types.

Can we get away with having `ProtocolParameters` in `Write` only? I would argue that the answer is "no". I completely agree that for making a transaction, we only need the latest protocol parameters, so that's a good argument in favor of "yes". However, the problem is that if the `cardano-node` is not fully synched, it might give us protocol parameters from old eras — and we need to be able to read those, e.g. for converting slot numbers to wall-clock time.

In that sense, it's fair game for `Read` to have a type `ProtocolParameters` for each era of the Cardano blockchain as they are defined in the specification. For instance, here is the protocol parameters specification for the Shelley era:

<img width="478" alt="PParams Shelley" src="https://user-images.githubusercontent.com/564592/231725507-39c4fe68-19cc-4c35-a2a8-641581e38413.png">

Reply:
In short, I'm actually in favor of *two* modules:

* `Cardano.Wallet.Read.ProtocolParameters`
* and `Cardano.Wallet.Write.ProtocolParameters`, which re-exports the previous one.

In order to preserve abstraction boundaries, modules in the (vicinity of the) `Write` hierarchy should import the `Cardano.Wallet.Write.ProtocolParameters`; this applies to many files in this pull request, see one comment.

Reply:
After discussing with Johannes, we came to the following conclusion:

* The type introduced in this pull request should live in `Cardano.Wallet.Write.ProtocolParameters`.

Furthermore:
* The `Cardano.Wallet.Write` hierarchy need not depend on `Cardano.Wallet.Read`, and can depend on ledger types directly.
  * I see `Cardano.Wallet.Read` as a way to represent concepts from the ledger specification of each era as Haskell types — i.e. I see it as a replacement for `cardano-ledger` (that also happens to be implemented using `cardano-ledger`). In this way, `Read` can absorb changes in the `cardano-ledger` code that are not reflected in the specification. This also implies that I find it natural to have `Write` depend on `Read`.
  * However, Johannes prefers `Write` to depend on `cardano-ledger` rather than `Read` because
    * Fee evaluation operates on types from `cardano-ledger`.
    * The dependency footprint for users of the `Write` hierarchy is reduced.
  * I don't object to the decision of having `Write` *not* depend on `Read` — as long as the abstraction "`Write` is about writing stuff in the current or subsequent era" is preserved, I would consider this dependency an implementation detail.
* We can't think of a good use case for `Cardano.Wallet.Read.ProtocolParameters` — old parameters may arise when the node is not fully synchronized, but the most "interesting" thing that the wallet can do in that case is to refuse making transactions.